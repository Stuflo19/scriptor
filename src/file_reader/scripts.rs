use std::collections::HashMap;

use crate::file_reader::read_json::read_json;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct Scripts {
    pub scripts: HashMap<String, String>,
}

impl Scripts {
    pub fn new() -> Self {
        let scripts = read_json().unwrap();

        Self { scripts }
    }

    // pub fn scripts(&self) -> &HashMap<String, String> {
    //     &self.scripts
    // }
}
