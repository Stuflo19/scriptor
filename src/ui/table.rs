use crate::file_reader::scripts::Scripts;
use ratatui::widgets::TableState;
use ratatui::{
    layout::Constraint,
    style::{Style, Stylize},
    text::{Span, Text},
    widgets::{Block, Cell, Row, Table},
};
use std::collections::HashMap;

pub struct ScriptTable {
    pub state: TableState,
    pub scripts: Scripts,
    pub filtered: HashMap<String, String>,
}

impl ScriptTable {
    pub fn new() -> Self {
        let scripts = Scripts::new();
        let temp = scripts.scripts.clone();

        Self {
            state: TableState::default().with_selected(0),
            scripts,
            filtered: temp,
        }
    }

    fn generate_table_header(&self) -> Row<'static> {
        let header_style = Style::default().on_black();

        ["Script", "Command"]
            .into_iter()
            .map(Cell::from)
            .collect::<Row>()
            .style(header_style)
            .height(1)
    }

    pub fn next(&mut self) {
        let i = match self.state.selected() {
            Some(i) => {
                if i >= self.scripts.scripts.len() - 1 {
                    0
                } else {
                    i + 1
                }
            }
            None => 0,
        };
        self.state.select(Some(i));
        //self.scroll_state = self.scroll_state.position(i * ITEM_HEIGHT);
    }

    pub fn previous(&mut self) {
        let i = match self.state.selected() {
            Some(i) => {
                if i == 0 {
                    self.scripts.scripts.len() - 1
                } else {
                    i - 1
                }
            }
            None => 0,
        };
        self.state.select(Some(i));
        //self.scroll_state = self.scroll_state.position(i * ITEM_HEIGHT);
    }

    pub fn filter_results(&mut self, input: String) {
        self.filtered = self
            .scripts
            .scripts
            .clone()
            .into_iter()
            .filter(|(key, _value)| key.contains(&input))
            .collect();
    }

    pub fn get_script_name(&self) -> String {
        let index = self.state.selected().unwrap();

        let keys: Vec<String> = self.scripts.scripts.clone().into_keys().collect();

        keys[index].clone()
    }

    fn generate_table_rows(&self) -> Vec<Row<'static>> {
        let script_rows = self
            .filtered
            .iter()
            .enumerate()
            .map(|(_i, (key, value))| {
                let content = [
                    Cell::from(Span::raw(format!("{key}"))),
                    Cell::from(Span::raw(format!("{value}"))),
                ];

                Row::new(content)
            })
            .collect::<Vec<Row>>();

        script_rows
    }

    pub fn generate_table(&self) -> Table<'static> {
        let bar = " â–ˆ ";

        let header = self.generate_table_header();

        let script_rows = self.generate_table_rows();

        let scripts_table = Table::new(
            script_rows,
            [
                Constraint::Length(25 + 1),
                Constraint::Min(25 + 1),
                Constraint::Min(25),
            ],
        )
        .header(header)
        .block(Block::bordered())
        .highlight_symbol(
            Text::from(vec![bar.into(), bar.into(), bar.into(), bar.into()])
                .style(Style::new().red()),
        );

        scripts_table
    }
}
