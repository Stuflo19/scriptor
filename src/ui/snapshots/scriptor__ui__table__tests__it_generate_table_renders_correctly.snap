---
source: src/ui/table.rs
expression: terminal.backend()
snapshot_kind: text
---
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"┌Scripts───────────────────────────────────────────────────────────────────────┐"
"│Script                    Command                                             │"
"│android                   yarn android                                        │"
"│ios                       yarn ios                                            │"
"│run                       expo start                                          │"
"│start                     yarn start                                          │"
"│test                      jest                                                │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"└──────────────────────────────────────────────────────────────────────────────┘"
